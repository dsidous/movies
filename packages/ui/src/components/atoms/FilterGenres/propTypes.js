import { array, func } from 'prop-types';

export const propTypes = {
  genres: array.isRequired,
  onChange: func.isRequired,
};

export const defaultProps = {};
