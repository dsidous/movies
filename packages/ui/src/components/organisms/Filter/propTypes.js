import { array, string, func, node } from 'prop-types';

export const propTypes = {
  query: string,
  media: string,
  queryUpdate: func,
  genres: array,
  children: node,
};

export const defaultProps = {};
