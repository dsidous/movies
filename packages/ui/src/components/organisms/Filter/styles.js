import { makeStyles } from '@material-ui/core/styles';

export default makeStyles(theme => ({
  'filter-wrapper': {
    maxWidth: '1150px',
    margin: '0 auto',

    '& form': {
      margin: `${theme.spacing(5)}px 0 ${theme.spacing(2)}px`,
    },
  },
}));
