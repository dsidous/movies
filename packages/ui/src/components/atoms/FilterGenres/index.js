import React, { useState, useEffect } from 'react';

import InputLabel from '@material-ui/core/InputLabel';
import MenuItem from '@material-ui/core/MenuItem';
import FormControl from '@material-ui/core/FormControl';
import Select from '@material-ui/core/Select';
import OutlinedInput from '@material-ui/core/OutlinedInput';

import { propTypes } from './propTypes';

const FilterGenres = ({ genres, onChange }) => {
  const [selectedGenres, setSelectedGenres] = useState({
    options: [],
    selectValue: [''],
  });

  useEffect(() => {
    const options = genres.map(genre => ({
      label: genre.name,
      value: String(genre.id),
    }));
    setSelectedGenres(s => ({ ...s, options }));
  }, [genres]);

  const handleChange = event => {
    setSelectedGenres({
      ...selectedGenres,
      selectValue: event.target.value,
    });
  };

  const handleClose = () => {
    onChange(selectedGenres.selectValue);
  };

  return (
    <FormControl className="filter-element-wrapper" variant="outlined">
      <InputLabel htmlFor="genres">Genres</InputLabel>
      <Select
        multiple
        value={selectedGenres.selectValue}
        onChange={handleChange}
        onClose={handleClose}
        input={<OutlinedInput id="genres" labelWidth={50} />}
        renderValue={selected => {
          if (selected.length === 1) {
            return <span>Select genres...</span>;
          }

          const list = selected
            .filter(select => select !== '')
            .map(
              select =>
                selectedGenres.options.find(option => option.value === select)
                  .label,
            );

          return list.join(', ');
        }}
      >
        <MenuItem value="" disabled>
          Select genres...
        </MenuItem>
        {selectedGenres.options.map(option => (
          <MenuItem key={option.label} value={option.value}>
            {option.label}
          </MenuItem>
        ))}
      </Select>
    </FormControl>
  );
};

FilterGenres.propTypes = propTypes;

export default FilterGenres;
