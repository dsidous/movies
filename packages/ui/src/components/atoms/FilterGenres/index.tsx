/** @jsxFrag React.Fragment */
import React, { useState, useEffect } from 'react';

import InputLabel from '@material-ui/core/InputLabel';
import MenuItem from '@material-ui/core/MenuItem';
import Select from '@material-ui/core/Select';
import OutlinedInput from '@material-ui/core/OutlinedInput';

import { propTypes } from './propTypes';

import { Genres } from '../../../types/movie';

type SelectValue = string;

type Option = {
  label: string;
  value: string;
};

interface Props {
  genres: Genres[];
  onChange: (arg0: SelectValue) => void;
}

interface Selected {
  options: Option[] | null;
  selectValue: SelectValue;
}

const FilterGenres: React.FC<Props> = ({ genres, onChange }: Props) => {
  const [selectedGenres, setSelectedGenres] = useState<Selected>({
    options: null,
    selectValue: '',
  });

  useEffect(() => {
    const options = genres.map(genre => ({
      label: genre.name,
      value: String(genre.id),
    }));
    setSelectedGenres(s => ({ ...s, options }));
  }, [genres]);

  const handleChange = (event: React.ChangeEvent<HTMLButtonElement>) => {
    setSelectedGenres({
      ...selectedGenres,
      selectValue: event.target.value,
    });
  };

  const handleClose = () => {
    onChange(selectedGenres.selectValue);
  };

  return (
    <>
      <InputLabel htmlFor="genres">Genres</InputLabel>
      <Select
        multiple
        value={selectedGenres.selectValue}
        onChange={handleChange}
        onClose={handleClose}
        input={<OutlinedInput id="genres" labelWidth={50} />}
        renderValue={(selected: []) => {
          if (!selected.length) {
            return <span>Select genres...</span>;
          }

          const list = selected
            .filter(select => select !== '')
            .map(select => {
              if (selectedGenres.options && selectedGenres.options.length > 0) {
                return selectedGenres?.options?.find(
                  (option: Option) => option.value === select,
                ).label;
              }

              return null;
            });

          return list.join(', ');
        }}
      >
        <MenuItem value="" disabled>
          Select genres...
        </MenuItem>
        {selectedGenres.options?.map(option => (
          <MenuItem key={option.label} value={option.value}>
            {option.label}
          </MenuItem>
        ))}
      </Select>
    </>
  );
};

FilterGenres.propTypes = propTypes;

export default FilterGenres;
