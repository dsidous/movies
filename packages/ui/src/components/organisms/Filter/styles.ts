import { makeStyles } from '@material-ui/core/styles';

export default makeStyles(
  theme => ({
    root: {
      margin: '0 auto',

      '& form': {
        display: 'grid',
        gridGap: theme.spacing(1),
        gridAutoFlow: 'column',
        margin: theme.spacing(5, 0, 2),
      },
    },
  }),
  { name: 'filter' },
);
