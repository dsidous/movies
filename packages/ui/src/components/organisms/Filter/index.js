/* eslint-disable camelcase */
import React, { useState, useEffect } from 'react';

import InputLabel from '@material-ui/core/InputLabel';
import MenuItem from '@material-ui/core/MenuItem';
import FormControl from '@material-ui/core/FormControl';
import Select from '@material-ui/core/Select';
import OutlinedInput from '@material-ui/core/OutlinedInput';

import { propTypes } from './propTypes';
import MyPager from '../../atoms/Pager';
import FilterGenres from '../../atoms/FilterGenres';
import useStyles from './styles';

const objectToQueryStr = paramsObj =>
  Object.keys(paramsObj)
    .map(
      key => `${encodeURIComponent(key)}=${encodeURIComponent(paramsObj[key])}`,
    )
    .join('&');

const initState = ({ params, release_year }) => ({
  page: parseInt(params.get('page'), 10) || 1,
  [release_year[0]]: params.get(release_year[0]) || new Date().getFullYear(),
  sort_by: params.get('sort_by') || 'popularity.desc',
  with_genres: params.get('with_genres') || '',
  'vote_average.gte': params.get('vote_average.gte') || 0,
});

const Filter = ({ query, media, genres, children, queryUpdate }) => {
  const classes = useStyles();
  const year = new Date().getFullYear();

  const params = new URLSearchParams(query);

  const release_year =
    media === 'tvs'
      ? ['first_air_date_year', 'first_air_date']
      : ['primary_release_year', 'primary_release_date'];

  const [state, setState] = useState(initState({ params, release_year }));

  useEffect(() => {
    const runQuery = () => {
      const strquery = objectToQueryStr(state);
      queryUpdate(strquery);
    };

    runQuery();
  }, [state, queryUpdate]);

  const sumYear = year - 1900;
  const yearOptions = Array(sumYear)
    .fill()
    .map((_, i) => (
      // eslint-disable-next-line react/no-array-index-key
      <MenuItem key={i} value={year - i}>
        {year - i}
      </MenuItem>
    ));

  const handleChange = e => {
    const val = e.target.value;
    const option = e.target.name;
    setState({ ...state, [option]: val, page: 1 });
  };

  const handleGenresChange = g =>
    setState({ ...state, with_genres: g, page: 1 });

  const handlePageSelect = page => {
    if (page > 0) {
      setState({ ...state, page });
    }
  };

  const { sort_by, with_genres, page } = state;
  return (
    <div className={classes['filter-wrapper']}>
      <form>
        <FormControl variant="outlined" className="filter-element-wrapper">
          <InputLabel htmlFor="shortby">Sort By</InputLabel>
          <Select
            id="shortby"
            placeholder="select"
            onChange={handleChange}
            value={sort_by}
            name="sort_by"
            input={<OutlinedInput labelWidth={53} id="shortby" />}
          >
            <MenuItem value="popularity.desc">Popularity Descending</MenuItem>
            <MenuItem value="popularity.asc">Popularity Ascending</MenuItem>
            <MenuItem value="vote_average.desc">Rating Descending</MenuItem>
            <MenuItem value="vote_average.asc">Rating Ascending</MenuItem>
            <MenuItem value={`${release_year[1]}.desc`}>
              Release Date Descending
            </MenuItem>
            <MenuItem value={`${release_year[1]}.asc`}>
              Release Date Ascending
            </MenuItem>
          </Select>
        </FormControl>
        <FormControl className="filter-element-wrapper" variant="outlined">
          <InputLabel>Year</InputLabel>
          <Select
            placeholder="select"
            onChange={handleChange}
            // eslint-disable-next-line react/destructuring-assignment
            value={state[release_year[0]]}
            name={release_year[0]}
            input={<OutlinedInput labelWidth={35} id="year" />}
          >
            <MenuItem value="null">None</MenuItem>
            {yearOptions}
          </Select>
        </FormControl>
        <FilterGenres
          genres={genres}
          onChange={handleGenresChange}
          value={with_genres}
        />
        <FormControl className="filter-element-wrapper" variant="outlined">
          <InputLabel>Average vote</InputLabel>
          <Select
            name="vote_average.gte"
            placeholder="select"
            onChange={handleChange}
            // eslint-disable-next-line react/destructuring-assignment
            value={state['vote_average.gte']}
            input={<OutlinedInput labelWidth={93} id="vote" />}
            fullWidth
          >
            <MenuItem value="0">Greater than...</MenuItem>
            <MenuItem value="9">9</MenuItem>
            <MenuItem value="8">8</MenuItem>
            <MenuItem value="7">7</MenuItem>
            <MenuItem value="6">6</MenuItem>
            <MenuItem value="5">5</MenuItem>
            <MenuItem value="4">4</MenuItem>
            <MenuItem value="3">3</MenuItem>
            <MenuItem value="2">2</MenuItem>
            <MenuItem value="1">1</MenuItem>
          </Select>
        </FormControl>
      </form>
      <MyPager page={page} handlePageSelect={handlePageSelect} />
      {children}

      <MyPager page={page} handlePageSelect={handlePageSelect} />
    </div>
  );
};

Filter.propTypes = propTypes;

export default Filter;
