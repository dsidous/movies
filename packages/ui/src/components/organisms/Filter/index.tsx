import React, { useState, useEffect, SetStateAction, Dispatch } from 'react';

import InputLabel from '@material-ui/core/InputLabel';
import MenuItem from '@material-ui/core/MenuItem';
import FormControl from '@material-ui/core/FormControl';
import Select from '@material-ui/core/Select';
import OutlinedInput from '@material-ui/core/OutlinedInput';

import MyPager from '../../atoms/Pager';
import FilterGenres from '../../atoms/FilterGenres';
import useStyles from './styles';
import { Genres } from '@typesRoots/movie';

interface Props {
  query: string;
  media: string;
  queryUpdate: Dispatch<SetStateAction<string>>;
  genres: Genres[];
  children: React.ReactNode;
}

interface InitState {
  params: URLSearchParams;
  release_year: string[];
}

interface ObjectToQueryStr {
  [x: string]: string | number;
  page: number;
  sort_by: string;
  with_genres: string;
  'vote_average.gte': string | number;
}

const objectToQueryStr = (paramsObj: ObjectToQueryStr) =>
  Object.keys(paramsObj)
    .map(
      (key) =>
        `${encodeURIComponent(key)}=${encodeURIComponent(paramsObj[key])}`
    )
    .join('&');

const initState = ({ params, release_year }: InitState) => ({
  page: +(params.get('page') || 1),
  [release_year[0]]: params.get(release_year[0]) || new Date().getFullYear(),
  sort_by: params.get('sort_by') || 'popularity.desc',
  with_genres: params.get('with_genres') || '',
  'vote_average.gte': params.get('vote_average.gte') || 0,
});

const Filter: React.FC<Props> = ({
  query,
  media,
  genres,
  children,
  queryUpdate,
}) => {
  const classes = useStyles();
  const year = new Date().getFullYear();

  const params = new URLSearchParams(query);

  const release_year =
    media === 'tvs'
      ? ['first_air_date_year', 'first_air_date']
      : ['primary_release_year', 'primary_release_date'];

  const [state, setState] = useState(initState({ params, release_year }));

  useEffect(() => {
    const runQuery = () => {
      const strquery = objectToQueryStr(state);
      queryUpdate(strquery);
    };

    runQuery();
  }, [state, queryUpdate]);

  const sumYear = year - 1900;
  const yearOptions = Array(sumYear)
    .fill(0)
    .map((_, i) => (
      // eslint-disable-next-line react/no-array-index-key
      <MenuItem key={i} value={year - i}>
        {year - i}
      </MenuItem>
    ));

  const handleChange = (e: React.ChangeEvent<HTMLSelectElement>) => {
    const val = e.target.value;
    const option = e.target.name;
    setState({ ...state, [option]: val, page: 1 });
  };

  const handleGenresChange = (g: string) =>
    setState({ ...state, with_genres: g, page: 1 });

  const handlePageSelect = (page: number) => {
    if (page > 0) {
      setState({ ...state, page });
    }
  };

  const { sort_by, with_genres, page } = state;
  return (
    <div className={classes.root}>
      <form>
        <FormControl variant='outlined'>
          <InputLabel htmlFor='shortby'>Sort By</InputLabel>
          <Select
            id='shortby'
            placeholder='select'
            onChange={handleChange}
            value={sort_by}
            name='sort_by'
            input={<OutlinedInput labelWidth={53} id='shortby' />}
          >
            <MenuItem value='popularity.desc'>Popularity Descending</MenuItem>
            <MenuItem value='popularity.asc'>Popularity Ascending</MenuItem>
            <MenuItem value='vote_average.desc'>Rating Descending</MenuItem>
            <MenuItem value='vote_average.asc'>Rating Ascending</MenuItem>
            <MenuItem value={`${release_year[1]}.desc`}>
              Release Date Descending
            </MenuItem>
            <MenuItem value={`${release_year[1]}.asc`}>
              Release Date Ascending
            </MenuItem>
          </Select>
        </FormControl>
        <FormControl variant='outlined'>
          <InputLabel>Year</InputLabel>
          <Select
            placeholder='select'
            onChange={handleChange}
            // eslint-disable-next-line react/destructuring-assignment
            value={state[release_year[0]]}
            name={release_year[0]}
            input={<OutlinedInput labelWidth={35} id='year' />}
          >
            <MenuItem value='null'>None</MenuItem>
            {yearOptions}
          </Select>
        </FormControl>
        <FormControl variant='outlined'>
          <FilterGenres genres={genres} onChange={handleGenresChange} />
        </FormControl>
        <FormControl variant='outlined'>
          <InputLabel>Average vote</InputLabel>
          <Select
            name='vote_average.gte'
            placeholder='select'
            onChange={handleChange}
            // eslint-disable-next-line react/destructuring-assignment
            value={state['vote_average.gte']}
            input={<OutlinedInput labelWidth={93} id='vote' />}
            fullWidth
          >
            <MenuItem value='0'>Greater than...</MenuItem>
            <MenuItem value='9'>9</MenuItem>
            <MenuItem value='8'>8</MenuItem>
            <MenuItem value='7'>7</MenuItem>
            <MenuItem value='6'>6</MenuItem>
            <MenuItem value='5'>5</MenuItem>
            <MenuItem value='4'>4</MenuItem>
            <MenuItem value='3'>3</MenuItem>
            <MenuItem value='2'>2</MenuItem>
            <MenuItem value='1'>1</MenuItem>
          </Select>
        </FormControl>
      </form>
      <MyPager page={page} handlePageSelect={handlePageSelect} />
      {children}

      <MyPager page={page} handlePageSelect={handlePageSelect} />
    </div>
  );
};

export default Filter;
