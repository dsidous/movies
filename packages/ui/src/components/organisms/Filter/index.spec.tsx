import React, { useState as useStateMock } from 'react';
import { shallow } from 'enzyme';

import Filter from '.';
import FilterGenres from '../../atoms/FilterGenres';
import MyPager from '../../atoms/Pager';

jest.mock('react', () => ({
  ...jest.requireActual('react'),
  useState: jest.fn(),
}));

const mockProps = {
  query: '',
  media: 'movies',
  queryUpdate: jest.fn(),
  genres: [
    { name: 'genre-1', id: '1' },
    { name: 'genre-2', id: '2' },
    { name: 'genre-3', id: '3' },
  ],
};

describe('Organisms/Filter', () => {
  let wrapper;
  const setState = jest.fn();

  beforeEach(() => {
    useStateMock.mockImplementation(init => [init, setState]);
    wrapper = shallow(
      <Filter {...mockProps}>
        <div />
      </Filter>,
    );
  });

  afterEach(() => {
    jest.clearAllMocks();
  });

  it('should render as expected', () => {
    expect(wrapper).toMatchSnapshot();
  });

  it('should handle genre select', () => {
    wrapper.find(FilterGenres).props().onChange('2');

    expect(setState).toHaveBeenCalledWith(
      expect.objectContaining({
        with_genres: '2',
      }),
    );
  });

  it('should handle page select', () => {
    wrapper.find(MyPager).first().props().handlePageSelect(2);

    expect(setState).toHaveBeenCalledWith(
      expect.objectContaining({
        page: 2,
      }),
    );
  });

  it('should handle "sort_by" select', () => {
    const selected = {
      target: { value: 'vote_average.desc', name: 'sort_by' },
    };

    wrapper.find('#shortby').props().onChange(selected);

    expect(setState).toHaveBeenCalledWith(
      expect.objectContaining({
        sort_by: selected.target.value,
      }),
    );
  });

  it('should handle "average vote" select', () => {
    const selected = { target: { value: '4', name: 'vote_average.gte' } };

    wrapper.find('[name="vote_average.gte"]').props().onChange(selected);

    expect(setState).toHaveBeenCalledWith(
      expect.objectContaining({
        'vote_average.gte': selected.target.value,
      }),
    );
  });
});
