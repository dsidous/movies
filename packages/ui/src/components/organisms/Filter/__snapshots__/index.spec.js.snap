// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Organisms/Filter should render as expected 1`] = `ShallowWrapper {}`;
